javac API change in Java 25 - Log.DiagnosticHandler

The internal javac `Log.DiagnosticHandler` API changes in Java 25: https://github.com/openjdk/jdk/commit/4890b74c048a1472b87687294c316ecfb324e4ba

This affects code in `javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java`

https://github.com/eisop/checker-framework/blob/5cc809d3f1c298b3ec12fa76e1f200fe5d1db2bf/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java#L220

The diagnostic handler implementations switch from being static member classes to non-static inner classes, so e.g. `new Log.DiscardDiagnosticHandler(log)` needs to change to `log. new DiscardDiagnosticHandler(log)`. Calling the constructor reflectively works with both versions, since either way it takes one argument for the log, whether that's the explicit parameter in Java < 25 or the synthetic enclosing instance parameter in Java >= 25.

Thanks for the heads-up!
In #1190 I'm improving our CI build to always run Gradle on JDK 21, which also allows us to already test JDK 25.
That build currently fails on JDK 25 with:

````
/home/runner/work/checker-framework/checker-framework/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java:265: error: an enclosing instance that contains Log.DiscardDiagnosticHandler is required
        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
                                                         ^
````

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -3,10 +3,14 @@ Version 3.49.1-eisop2 (March ??, 2025)
 
 **User-visible changes:**
 
+The Checker Framework runs under JDK 25 -- that is, it runs on a version 25 JVM.
+
 **Implementation details:**
 
 **Closed issues:**
-eisop#1051
+
+eisop#1051, eisop#1180.
+
 
 Version 3.49.1-eisop1 (March 17, 2025)
 --------------------------------------

--- a/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java
@@ -209,6 +209,23 @@ public class Resolver {
         }
     }
 
+    protected static Log.DiagnosticHandler createDiscardDiagnosticHandler(Log log) {
+        Constructor<Log.DiscardDiagnosticHandler> cons;
+        try {
+            cons = Log.DiscardDiagnosticHandler.class.getConstructor(Log.class);
+        } catch (NoSuchMethodException nsme) {
+            throw new BugInCF("Could not find Log.DiscardDiagnosticHandler constructor");
+        }
+        Log.DiagnosticHandler res;
+        try {
+            res = cons.newInstance(log);
+        } catch (IllegalAccessException | InstantiationException | InvocationTargetException e) {
+            throw new BugInCF(
+                    "Exception when invoking Log.DiscardDiagnosticHandler constructor", e);
+        }
+        return res;
+    }
+
     /**
      * Finds the package with name {@code name}.
      *
@@ -217,7 +234,7 @@ public class Resolver {
      * @return the {@code PackageSymbol} for the package if it is found, {@code null} otherwise
      */
     public @Nullable PackageSymbol findPackage(String name, TreePath path) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
             final Element res;
@@ -262,7 +279,7 @@ public class Resolver {
      * @return the element for the field, {@code null} otherwise
      */
     public @Nullable VariableElement findField(String name, TypeMirror type, TreePath path) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
             final Element res;
@@ -308,7 +325,7 @@ public class Resolver {
      * @return the element for the local variable, {@code null} otherwise
      */
     public @Nullable VariableElement findLocalVariableOrParameter(String name, TreePath path) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
             // Either a VariableElement or a SymbolNotFoundError.
@@ -356,7 +373,7 @@ public class Resolver {
      * @return the element for the class
      */
     public Element findClass(String name, TreePath path) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
             return wrapInvocationOnResolveInstance(FIND_TYPE, env, names.fromString(name));
@@ -374,7 +391,7 @@ public class Resolver {
      * @return the {@code ClassSymbol} for the class if it is found, {@code null} otherwise
      */
     public @Nullable ClassSymbol findClassInPackage(String name, PackageSymbol pck, TreePath path) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
             final Element res;
@@ -428,7 +445,7 @@ public class Resolver {
             TypeMirror receiverType,
             TreePath path,
             java.util.List<TypeMirror> argumentTypes) {
-        Log.DiagnosticHandler discardDiagnosticHandler = new Log.DiscardDiagnosticHandler(log);
+        Log.DiagnosticHandler discardDiagnosticHandler = createDiscardDiagnosticHandler(log);
         try {
             Env<AttrContext> env = getEnvForPath(path);
 

--- a/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/Resolver.java
@@ -209,6 +209,14 @@ public class Resolver {
         }
     }
 
+    /**
+     * Reflectively create a `Log.DiagnosticHandler`. Java 25 changed this class from a static
+     * member class to a non-static inner class. Reflectively creating an instance is easy, however,
+     * as both take the same arguments.
+     *
+     * @param log the (enclosing) Log instance
+     * @return a new Log.DiscardDiagnosticHandler
+     */
     protected static Log.DiagnosticHandler createDiscardDiagnosticHandler(Log log) {
         Constructor<Log.DiscardDiagnosticHandler> cons;
         try {

